### ISSUE TITLE
Struggling with YOLO setup – First attempts, but stuck

### ISSUE THREAD

**Issue by Rexem99** [MultimediaTechLab/YOLO#190]
Hi guys,

I recently started working with computer vision as part of my thesis. Initially, I achieved good results using the Ultralytics YOLO model, but its license poses a challenge for future use cases. After some research, I believe this YOLO repository is the best fit for my needs. I also tried YOLOX, DAMO-YOLO, and YOLO-NAS but couldn’t get them to work.

I now have some questions and problems regarding the training process and on how to correctly load a trained model for predictions. I would be very happy if someone can give me some ideas on how to solve these problems. 

I am training on a custom dataset in the normalized yolo labeling format (c,x,y,w,h) with the following folder structure and configs:
#######################################
dataset/
├── images
│   ├── train
│   └── val
├── labels
│   ├── train
│   └── val
└── classes.txt
#######################################
dataset.yaml:
  path: dataset_dog/
  train: train
  validation: val
  
  class_num: 4
  class_list: ["Bicycle","Car","Cat","Dog"]
#######################################
config.yaml:

    hydra:
      run:
        dir: runs
    
    name: v9-train-dog
    
    defaults:
      - _self_
      - task: train
      - dataset: yolo_dog
      - model: v9-m
      - general
#######################################
general.yaml:

    device: cuda
    cpu_num: 8
    
    image_size: [640, 640]
    
    out_path: runs
    exist_ok: True
    
    lucky_number: 10
    use_wandb: False
    use_tensorboard: True
    
    weight: yolo/v9-m.pt
#######################################

1.  While training on my custom dataset the training metrics look like this:
![Image](https://github.com/user-attachments/assets/fc45929d-fedf-4bd3-804d-9847005c0030)

2. After training the last checkpoint is being saved in the runs folder. I am currently loading the model how it is suggested in the lazy.py and I do not get any predictions on my pictures. I am not sure if I am doing something wrong with the training or the prediction and on how to validate if the model was trained correctly or not.

The configs for predictions look like this:
#######################################
config_inference.yaml

    hydra:
      run:
        dir: runs
    
    name: v9-inf
    
    defaults:
      - _self_
      - task: inference
      - dataset: yolo_dog
      - model: v9-m 
      - general_inference
#######################################
general_inference.yaml

    device: cuda
    cpu_num: 8
    
    image_size: [640, 640]
    
    out_path: runs
    exist_ok: True
    
    lucky_number: 10
    use_wandb: False
    use_tensorboard: True

    weight: yolo/v9-m-custom.ckpt
#######################################

3. 
What is the correct way to load the model for predictions? I want to setup a pipeline in which I can predict on camera (not webcam) pictures in real time inside a loop with images in opencv format and get the results to work with (which class was predictied etc. like in ultralytics).

**Comment by divyanidhi** [comment]
Hi @Rexem99 
Were you able to solve this issue?

**Comment by divyanidhi** [comment]
My training data is in YOLO detection format (cx, cy, w, h). While training, since I had incorporated #PR175,  it converts bounding boxes to YOLO segmentation format  if they are in YOLO detection format.
I had trained a model with 30 epochs. Loss looks something like this:

![Image](https://github.com/user-attachments/assets/6f6e8cf1-b108-4722-9615-82211f4dd6a4)

I am using below code for inferencing. 

!python yolo/lazy.py task=inference model=v9-c device=gpu weight=custom-trained-model task.nms.min_confidence=0.1 task.data.source=test_image.png use_wandb=False

**Comment by fjodborg** [comment]
Hmm, to me the training looks alright. But the output image puts all my suspicion on the the bounding box format somewhere not being correct. 
Again due to the bounding box having its corner in the center of a correct object. Which to me seems like a cx and cy is being interpreted as a corner instead of a center. 

I can't think of other things, sorry.

**Comment by fjodborg** [comment]
I'm not the author of the repo and it isn't a plug and play repo, on that we can agree. But honestly, all of it has most likely been freetime work and people need to understand that. If i maintained an open source project for fun i would most likely also focus on the fun parts, which usually isn't the documentation or small edge cases. They would probably be happy if someone volunteered to help maintain it, but few do so. 

Honestly, just be thankful that all the hardwork is already complete in this repo and only a few minor documentation things are missing, which can be derived if you read through the code or the open pull requests. They made it open source under MIT license, so if you aren't happy, just fork it and make your changes or perhaps contribute to the repo? Maybe you could find a fork that has better documentation, or if you really want something plug and play, perhaps use ultralyics and pay for their commercial license if you need that.

**Comment by pzoltowski** [comment]
I know. It's just hard to even contribute to docs o repo if someone cannot even test it. Half of issues on this repo are just marked as questions

**Comment by pdisbeschl** [comment]
> Hi [@fjodborg](https://github.com/fjodborg)
> 
> Adding this PR to code does solve the issue of yolo format, but I am not able to get correct inferencing on the custom trained model. Adding image as a reference.
> 
> ![Image](https://github.com/user-attachments/assets/3adfad54-30c2-4e8a-a092-264f5dacea63)
> 
> Do you have any suggestions on what could be the issue?

Hi @divyanidhi, thanks for the pointer! I managed to solve the issue in my case (it's related to what's discussed in https://github.com/MultimediaTechLab/YOLO/pull/175). 
In [solver.py](https://github.com/MultimediaTechLab/YOLO/blob/main/yolo/tools/solver.py), the `InferenceModel.predict_step` function retrieves the `predicts` containing the locations of the bounding boxes.

I can't show my picture, but these are the "predicts" from YOLO that I get: 
`[0 48.3438 21.3750 142.8750 81.5000 0.9697]`
This translates to:
[`class id`  `width`  `height`  `center_width`  `center_height`  `accuracy`], where the dimensions are **pixel values**!
So if you want to get the bounding box for this object:

> `left_top_corner = center_width-0.5*width + center_height-0.5*height`
> `right_top_corner = center_width+0.5*width + center_height-0.5*height`
> `right_bottom_corner = center_width+0.5*width + center_height+0.5*height`
> `left_bottom_corner = center_width-0.5*width + center_height+0.5*height`
> `bbox = [left_top_corner, right_top_corner, right_bottom_corner, left_bottom_corner]`

These values are different from what I provided as the training input for my model:
[0 0.4434 0.3412 0.1565 0.0917]
[`class_id`, `center_width`, `center_height`, `width`, `height`]

--> *These are quotients corresponding to the pixel locations of images divided by the image's width and height, in pixels.*

**NOTE:** I have not tried anybody else's proposed fixes yet, I might create a pull request to solve this issue.
